### ISSUE TITLE
[HighLevelRest] UpdateRequest with Script works incorrectly

### ISSUE THREAD

**Issue by casper1149** [elastic/elasticsearch#33221]
Hello,

Elasticsearch version: 6.4.0

Documentation: https://www.elastic.co/guide/en/elasticsearch/client/java-rest/6.4/java-rest-high-document-update.html#java-rest-high-document-update

JVM version: 1.8.0

OS version: MacOSX 17.6.0 Darwin Kernel

I have an index with next mapping
```
{
  "_doc": {
    "properties": {
      "steps": {
        "type": "nested"
      }
    }
  }
}
```

I indexed a simple document
```
PUT myindex/_doc/1
{
  "steps": []
}
```

In order to update the "steps" array I created next query
```
POST myindex/_doc/1/_update
{
  "script": {
    "source": "ctx._source.steps.add(params.step)",
    "lang": "painless",
    "params": {
      "step": {
        "field1": {
          "field2": "value",
          "field3": "value"
        },
        "field4": "value",
      }
    }
  }
}
```

And it works just fine, however when I try to do the same using the Java High Level Rest Client, something goes wrong, here is my code
```
JsonNode jsonNode = objectMapper.convertValue(step, JsonNode.class);
UpdateRequest updateRequest = new UpdateRequest()
    .index("myindex")
    .type("_doc")
    .id("1")
    .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
    .script(new Script(ScriptType.INLINE, "painless",
        "ctx._source.steps.add(params.step)", singletonMap("step", jsonNode)));
client.update(updateRequest, RequestOptions.DEFAULT);
```

the "jsonNode" looks like this when I print it
```
{
"field4":"value",
"field1":{"field2":"value","field3":"value"}
}
```

the "singletonMap" also looks correct

As a result in Kibana I can see next result
```
          "steps": [
            [
              [],
              [
                [],
                []
              ]
            ]
           ]
```

**Comment by casper1149** [comment]
Just to provide more information, I tried to use the low level client and it works as expected
```
        String json = String.format("{\n" +
                        "  \"script\": {\n" +
                        "    \"source\": \"ctx._source.steps.add(params.step)\",\n" +
                        "    \"lang\": \"painless\",\n" +
                        "    \"params\": {\n" +
                        "      \"step\": %s\n" +
                        "    }\n" +
                        "  }\n" +
                        "}",
                objectMapper.writeValueAsString(step));

        Request request = new Request("POST", "<url>/myindex/_doc/1/_update");
        request.addParameter("refresh", "true");
        request.setEntity(new NStringEntity(json, ContentType.APPLICATION_JSON));
        client.getLowLevelClient().performRequest(request);
```

**Comment by hub-cap** [comment]
Hi,

First off, thank you for a great report and reproduction steps. I went ahead and built out a quick test for this, and found out what you are running up against. 

A JsonNode is Iterable. Our XContentBuilder is assuming that any Iterable is an array in [json](https://github.com/elastic/elasticsearch/blob/master/libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java#L834-L839), which in turn is doing the [following](https://github.com/codehaus/jackson/blob/master/src/java/org/codehaus/jackson/JsonNode.java#L747-L761), which ends up parsing in to something that you dont want it to. 

Your best bet here is to first convert the jsonNode to a Map. One way I found to do it was by using the ObjectMapper `Map jsonMap = mapper.convertValue(jsonNode, Map.class)` and pass that in to the `singletonMap("step", jsonAsMap)`. I have verified this ends up giving the following

```
 curl localhost:9200/myindex/_doc/1
{"_index":"myindex","_type":"_doc","_id":"1","_version":2,"found":true,"_source":{"steps":[{"field1":{"field3":"value","field2":"value"},"field4":"value"}]}}
```

Give that a try to see if it solves your issue.
